### PR TITLE
Prevent opening vending UI multiple times

### LINKED ISSUE #3408: Autosave prevent vending when save happen while setting up shop
* **rAthena Hash**: 3fc88fe6c

* **Client Date**: 20151104 / 20180530

* **Server Mode**: Pre-RE

* **Description of Issue**: 
  * Result: When an autosave happen when setting up a shop, you won't be able to complete it (video of the issue: https://mega.nz/#!hJ8ixYxL!RNV1nqkosGYajQJyb7Qs2vPI0oDWRSJExrx1IQL312U)
Meaning that when you set autosave_time is too low, players will be unable to setup shops, or even if the value is high, if you have back luck, autosave will happen as soon as you start making the shop and you won't be able to complete it.
  * Expected Result: You should be able to setup you shop even if you take time to do it
  * How to Reproduce: 
-Start opening a shop
-Wait for an autosave
-Try to confirm shop creation, you won't be able to
  * Official Information:/
* **Modifications that may affect results**: 
the lower the autosave_time is, the lower the chance of the issue happening

### PR DESCRIPTION
**Pull request by secretdataz**
* **Addressed Issue(s)**: Fixes #3408 

* **Server Mode**: Both

* **Description of Pull Request**: No vending open packet (ZC_OPENSTORE) will be sent on autosave if the vending UI is already open to prevent vending failure.

### CODE REVIEW

**Comment by Lelouzh** [comment]
Confirm. Everytime autosave time starting save when making Vending. The Vending is bug and need to re-open again. After trying this code. I can hold on Vending more than char/map auto save times

GOOD WORK

File: src/map/pc.hpp [hunk 1]
```diff
@@ -364,6 +364,7 @@ struct map_session_data {
 		unsigned int autolooting : 1; //performance-saver, autolooting state for @alootid
 		unsigned int gmaster_flag : 1;
 		unsigned int prevend : 1;//used to flag wheather you've spent 40sp to open the vending or not.
+		unsigned int pending_vending_ui : 1; // flag whether this player still needs
```
Comment (Lemongrass3110) [inline src/map/pc.hpp:367]: Looks like your comment was cut off?

File: src/map/intif.cpp [hunk 2]
```diff
@@ -3514,7 +3514,10 @@ static bool intif_parse_StorageReceived(int fd)
 			}else if( sd->state.prevend ){
 				clif_clearcart(sd->fd);
 				clif_cartlist(sd);
-				clif_openvendingreq(sd, sd->vend_skill_lv+2);
+				if (sd->state.pending_vending_ui) {
```
Comment (Lemongrass3110) [inline src/map/intif.cpp:3517]: Maybe add a little comment why we do it like this.

File: src/map/clif.cpp [hunk 3]
```diff
@@ -12368,7 +12366,7 @@ void clif_parse_PutItemToCart(int fd,struct map_session_data *sd)
 void clif_parse_GetItemFromCart(int fd,struct map_session_data *sd)
 {
 	struct s_packet_db* info = &packet_db[RFIFOW(fd,0)];
-	if (!pc_iscarton(sd))
+	if (!pc_iscarton(sd) || pc_cant_act2(sd))
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:12369]: There is a bunch of checks inside pc_getitemfromcart.
Maybe check if you can move those two checks into the other function?

File: src/map/clif.cpp [hunk 4]
```diff
@@ -12355,9 +12355,7 @@ void clif_parse_StopAttack(int fd,struct map_session_data *sd)
 void clif_parse_PutItemToCart(int fd,struct map_session_data *sd)
 {
 	struct s_packet_db* info = &packet_db[RFIFOW(fd,0)];
-	if (pc_istrading(sd))
-		return;
-	if (!pc_iscarton(sd))
+	if (pc_istrading(sd) || !pc_iscarton(sd) || pc_cant_act2(sd))
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:12358]: There is a bunch of checks inside pc_putitemtocart.
Maybe check if you can move those two checks into the other function?

File: src/map/intif.cpp [hunk 5]
```diff
@@ -3514,7 +3514,10 @@ static bool intif_parse_StorageReceived(int fd)
...
+					clif_openvendingreq(sd, sd->vend_skill_lv + 2);
+					sd->state.pending_vending_ui = 0;
```
Comment (Lemongrass3110) [inline src/map/intif.cpp:3519]:
```suggestion
					sd->state.pending_vending_ui = false;
```

File: src/map/skill.cpp [hunk 6]
```diff
@@ -8156,8 +8156,10 @@ int skill_castend_nodamage_id (struct block_list *src, struct block_list *bl, ui
 				sd->state.workinprogress = WIP_DISABLE_ALL;
 				sd->vend_skill_lv = skill_lv;
 				ARR_FIND(0, MAX_CART, i, sd->cart.u.items_cart[i].nameid && sd->cart.u.items_cart[i].id == 0);
-				if (i < MAX_CART)
+				if (i < MAX_CART) {
+					sd->state.pending_vending_ui = 1;
```
Comment (Lemongrass3110) [inline src/map/skill.cpp:8160]:
```suggestion
					sd->state.pending_vending_ui = true;
```

**Comment by zdurexz** [comment]
Wait complete and merge to master

**Review by Lemongrass3110 — APPROVED** [review]

**Review by aleos89 — APPROVED** [review]

**Pull request merged**